cross-chain/bob/contracts/OptimismMintableUpgradableTBTC.sol

```diff
@@ -0,0 +1,337 @@
+///           avoidance of contagion in case of a chain- or bridge-specific
+///           incident.
+///         - Be governed and upgradeable.
+///
+///         The token is burnable by the token holder and supports EIP2612
+///         permits. Token holder can authorize a transfer of their token with
+///         a signature conforming EIP712 standard instead of an on-chain
+///         transaction from their address. Anyone can submit this signature on
+///         the user's behalf by calling the permit function, paying gas fees,
+///         and possibly performing other actions in the same transaction.
+///         The governance can recover ERC20 and ERC721 tokens sent mistakenly
+///         to OptimismMintableUpgradableTBTC token contract.
+/// @custom:oz-upgrades-from OptimismMintableUpgradableERC20
+contract OptimismMintableUpgradableTBTC is
+    OptimismMintableUpgradableERC20,
+    OwnableUpgradeable,
+    PausableUpgradeable
+{
+    using SafeERC20Upgradeable for IERC20Upgradeable;
+
+    /// @notice Indicates if the given address is a minter. Only minters can
+    ///         mint the token.
+    mapping(address => bool) public isMinter;
+
+    /// @notice List of all minters.
+    address[] public minters;
+
+    /// @notice Indicates if the given address is a guardian. Only guardians can
+    ///         pause token mints and burns.
+    mapping(address => bool) public isGuardian;
+
+    /// @notice List of all guardians.
+    address[] public guardians;
+
+    /// @notice Tracks the remaining capacity for legacy bridge operations.
+    ///         Incremented on bridge mints, decremented on bridge burns.
+    uint256 public legacyCapRemaining;
+
+    event MinterAdded(address indexed minter);
+    event MinterRemoved(address indexed minter);
+
+    event GuardianAdded(address indexed guardian);
+    event GuardianRemoved(address indexed guardian);
+
+    /// @dev Modifier to make a function callable only when the contract is not paused.
+    /// @dev Requirements:
+    ///      - The contract must not be paused.
+    modifier onlyMinter() {
+        require(isMinter[msg.sender], "Caller is not a minter");
+        _;
+    }
+
+    /// @dev Modifier to make a function callable only when the contract is not paused.
+    /// @dev Requirements:
+    ///      - The contract must not be paused.
+    modifier onlyGuardian() {
+        require(isGuardian[msg.sender], "Caller is not a guardian");
+        _;
+    }
+
+    /// @custom:oz-upgrades-unsafe-allow constructor
+    constructor() {
+        _disableInitializers();
+    }
+
+    /// @notice Initialize v2 of the contract with new functionality
+    /// @custom:oz-upgrades-validate-as-initializer
+    function initializeV2() public reinitializer(2) {
+        __Ownable_init();
+        __Pausable_init();
+
+        // Set legacyCapRemaining to current total supply
+        legacyCapRemaining = totalSupply();
```

Comment:
can we document briefly a bit more the Legacy Cap mechanism? Is not 100% clear for me, I can imagine how it works, probably for others the same.

Reply:
Document added in [chore: add bob upgrade technical doc](https://github.com/threshold-network/tbtc-v2/pull/888/commits/0416da8dcbe1c7a2612f153d13d85251914f3b83)